File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        c = 0
        m = len(grid)
        n = len(grid[0])
        def makeAdjZero(i, j):
            if i < 0 or i >= m or j < 0 or j >= n:
                return
            if grid[i][j] == "0":
                return
            grid[i][j] = "0"
            makeAdjZero(i, j+1)
            makeAdjZero(i, j-1)
            makeAdjZero(i+1, j)
            makeAdjZero(i-1, j)
            return
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    c += 1
                    makeAdjZero(i, j)
        return c
                
            

File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int count = 0;
        int[][] dir = {{1,0}, {0,-1}, {-1, 0}, {0, 1}};
        for(int i = 0; i<m; i++) {
            for(int j = 0; j<n; j++) {
                if(grid[i][j] == '1') {
                    count += 1;
                    dfs(grid, i, j, dir);
                }
            }
        }
        return count;
    }
    public void dfs(char[][] grid, int i, int j, int[][] dir) {
        int m = grid.length;
        int n = grid[0].length;
        if(i < 0 || j < 0 || i >= m || j >= n || grid[i][j] == '0' || grid[i][j] == '2') {
            return;
        }
        grid[i][j] = '2';
        for(int[] d : dir) {
            int nexti = i + d[0];
            int nextj = j + d[1];
            dfs(grid, nexti, nextj, dir);
        }
    }
}